rust/flexbuffers/src/reader/mod.rs

```diff
@@ -266,65 +269,89 @@ impl<'de> Reader<'de> {
             })
         }
     }
-    /// Directly reads a slice of type `T`where `T` is one of `u8,u16,u32,u64,i8,i16,i32,i64,f32,f64`.
+
+    /// Directly reads a slice of type `T` where `T` is one of `u8,u16,u32,u64,i8,i16,i32,i64,f32,f64`.
     /// Returns Err if the type, bitwidth, or memory alignment does not match. Since the bitwidth is
     /// dynamic, its better to use a VectorReader unless you know your data and performance is critical.
     #[cfg(target_endian = "little")]
-    pub fn get_slice<T: ReadLE>(&self) -> Result<&'de [T], Error> {
+    pub fn get_slice<T: ReadLE>(&self) -> Result<&[T], Error> {
         if self.flexbuffer_type().typed_vector_type() != T::VECTOR_TYPE.typed_vector_type() {
             self.expect_type(T::VECTOR_TYPE)?;
         }
         if self.bitwidth().n_bytes() != std::mem::size_of::<T>() {
             self.expect_bw(T::WIDTH)?;
         }
         let end = self.address + self.length() * std::mem::size_of::<T>();
-        let slice = &self
+        let slice: &[u8] = &self
             .buffer
+            .as_ref()
             .get(self.address..end)
             .ok_or(Error::FlexbufferOutOfBounds)?;
+
         // `align_to` is required because the point of this function is to directly hand back a
         // slice of scalars. This can fail because Rust's default allocator is not 16byte aligned
         // (though in practice this only happens for small buffers).
-        let (pre, mid, suf) = unsafe { slice.align_to::<T>() };
+        let (pre, mid, suf) = unsafe { slice.as_ref().align_to::<T>() };
         if pre.is_empty() && suf.is_empty() {
             Ok(mid)
         } else {
             Err(Error::AlignmentError)
         }
     }
 
+    /// Returns the value of the reader if it is a boolean.
+    /// Otherwise Returns error.
     pub fn get_bool(&self) -> Result<bool, Error> {
         self.expect_type(FlexBufferType::Bool)?;
         Ok(
-            self.buffer[self.address..self.address + self.width.n_bytes()]
+            self.buffer.as_ref()[self.address..self.address + self.width.n_bytes()]
                 .iter()
                 .any(|&b| b != 0),
         )
     }
-    pub fn get_key(&self) -> Result<&'de str, Error> {
+
+    /// Gets the length of the key if this type is a key.
+    ///
+    /// Otherwise, returns an error.
+    fn get_key_len(&self) -> Result<usize, Error> {
         self.expect_type(FlexBufferType::Key)?;
-        let (length, _) = self.buffer[self.address..]
+        let (length, _) = self.buffer.as_ref()[self.address..]
             .iter()
             .enumerate()
             .find(|(_, &b)| b == b'\0')
             .unwrap_or((0, &0));
-        let bytes = &self.buffer[self.address..self.address + length];
+        Ok(length)
+    }
+
+    fn get_key_bytes(&self) -> Result<B, Error> {
```

Comment:
I think the problem that you're fighting is that we return things that maybe outlive the `Reader`, depending if its backed by `&'de [u8]` vs `Bytes`. These come out as `&'de str` and `&'de [u8]` and are seen in the key iterator and blob stuff.

Idea (not necessarily good) what if you made the return types of `get_str` and `get_blob`, and the iterator item of of `iter_keys` associated types of `InternalBuffer`? 

I suspect that Rust cannot support easily the kind of generics we want here unfortunately and it seems like what I did relies on the data being owned outside Reader quite a lot

Reply:
I think that could work. 

Alternatively I was thinking I could implement certain methods specifically for the `&'de [u8]` type, then provide an `AsRef<Reader<[u8]>>` implementation for `Reader<B>`. 

Something like this - 

```rust
impl<'de> Reader<&'de [u8]> {
    /// Retrieves str pointer, errors if invalid UTF-8, or the provided index
    /// is out of bounds.
    pub fn borrow_get_str(&self) -> Result<&'de str, Error> {
        self.expect_type(FlexBufferType::String)?;
        let bytes = self.buffer.as_ref().get(self.address..self.address + self.length());
        Ok(std::str::from_utf8(bytes.ok_or(Error::ReadUsizeOverflowed)?)?)
    }

    /// Returns empty string if you're not trying to read a string.
    pub fn borrow_as_str(&self) -> &'de str {
        match self.fxb_type {
            FlexBufferType::String => self.borrow_get_str().unwrap_or_default(),
            FlexBufferType::Key => self.borrow_get_key().unwrap_or_default(),
            _ => "",
        }
    }

    /// borrows a string provided by the get_key method. 
    pub fn borrow_get_key(&self) -> Result<&'de str, Error> {
        let bytes = &self.buffer.as_ref()[self.address..self.address + self.get_key_len()?];
        Ok(std::str::from_utf8(bytes)?)
    }
}

impl<B: InternalBuffer> AsRef<Reader<&[u8]>> for Reader<B> {
    fn as_ref(&self) -> Reader<&[u8]> {
        Reader {
            fxb_type: self.fxb_type,
            width: self.width,
            address: self.address,
            buffer: self.buffer.as_ref(),
        }
    }
}
```

It's still not ideal as most use cases will be required to use &[u8] references. However now users will be able to pass readers over channels etc without worry about lifetimes (i.e. via `Bytes`), then simply call `as_ref` if they need to to pass `&str`s around. 

I could add something similar for `MapReader` as well.

Reply:
Welp, unfortunately that AsRef impl gives me a weird error - 

```
[rustc] `impl` item signature doesn't match `trait` item signature

found `fn(&reader::Reader<B>) -> &reader::Reader<&[u8]>`

note: expected `fn(&reader::Reader<B>) -> &reader::Reader<&[u8]>`
         found `fn(&reader::Reader<B>) -> &reader::Reader<&[u8]>`
help: the lifetime requirements from the `impl` do not correspond to the requirements in the `trait`
help: verify the lifetime relationships in the `trait` and `impl` between the `self` argument, the other inputs and its output
```

Reply:
hmmm yea ultimately I don't think this will work out since we're trying to abstract between these very different functions

```rust
fn get_borrowed_bytes(&'a self) -> &'de [u8];  // requires bytes to exist at lifetime 'de which outlives self
fn get_owned_bytes(&'self self) -> &'self [u8];
```

One way to make this work is to sacrifice backwards compatibility... disallow `Reader<B>` from returning a reference that outlives it (indeed it already cannot due to the lifetime incompatibility of `AsRef::as_ref` that I mentioned earlier), and add methods like `get_borrowed_str` to `impl Reader<&'de[u8]>` specifically, which existing users may migrate to. (Maybe "borrowed" isn't the right word, its like `get_str_with_lifetime_from_underlying_buffer`)

though, I'm not fully convinced that this is worth breaking compatibility or the complexity

Reply:
Hm, would it be easier to just implement `get_str`, `as_str` and `get_key` for the `Reader<&'de [u8]>` type, rather than having the dupe functionality between `get_str` and `borrow_get_str` etc. 

That way the current functionality will work as expected, and if another user in the future wants to create / pass around a `Reader<B>`, they'll still be able to. 

I feel that could be the "best of both worlds" solution. 

Also, let me know if you'd like to discuss this on Discord for some clarification / discussion, I have the same tag (colindjk) in the `FlatBuffers` discord.

Reply:
I also wanted to add, if this PR is to go through, adding mutability to FlexBuffers wouldn't be too difficult of a next step. 

Basically just adding a InternalBufMut trait which could modify fields of the same size (plus maybe some restrictions through the Reader field based on FlexBufferType).